Jam/4525 globs file

JordanMartinez opened the pull request:
**Description of the change**

Fixes #4525.

---

**Checklist:**

- [x] Added a file to CHANGELOG.d for this PR (see CHANGELOG.d/README.md)
- [ ] Added myself to CONTRIBUTORS.md (if this is my first contribution)
- [x] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [ ] Added a test for the contribution (if applicable)

JordanMartinez commented:
`glob-test.sh` outputs the following:
```sh
$ stack exec bash glob-test.sh 
env info
0.15.14 [development build; commit: e25c476c08c2e134f5d369326060be2f0d9ef583 DIRTY]
z
├── Foo.purs
├── globsAll
├── globsNoFoo
└── src
    ├── Bar
    │   └── Baz.purs
    └── Bar.purs

2 directories, 5 files
---
module Foo where
foo = 1 :: Int
---
module Bar where
bar = 1 :: Int
---
module Bar.Baz where
baz = 1 :: Int
---
Result
z/out1
├── Bar.Baz
│   ├── externs.cbor
│   └── index.js
├── cache-db.json
├── Foo
│   ├── externs.cbor
│   └── index.js
└── package.json

2 directories, 6 files
z/out2
├── Bar
│   ├── externs.cbor
│   └── index.js
├── Bar.Baz
│   ├── externs.cbor
│   └── index.js
├── cache-db.json
├── Foo
│   ├── externs.cbor
│   └── index.js
└── package.json

3 directories, 8 files
z/out3
├── Bar
│   ├── externs.cbor
│   └── index.js
├── Bar.Baz
│   ├── externs.cbor
│   └── index.js
├── cache-db.json
├── Foo
│   ├── externs.cbor
│   └── index.js
└── package.json

3 directories, 8 files
z/out4
├── Bar
│   ├── externs.cbor
│   └── index.js
├── Bar.Baz
│   ├── externs.cbor
│   └── index.js
├── cache-db.json
├── Foo
│   ├── externs.cbor
│   └── index.js
└── package.json

3 directories, 8 files
SOURCE_GLOBS output different from EXPECTED
Expected: purs compile: No files found using pattern: z/test/**/*.purs
[1 of 2] Compiling Bar.Baz
[2 of 2] Compiling Foo
SOURCE_GLOBS: purs compile: No files found using pattern: z/test/**/*.purs
[1 of 3] Compiling Foo
[2 of 3] Compiling Bar
[3 of 3] Compiling Bar.Baz
```

I tried this script on `master` and the EXPECTED globs also exclude the `[2 of 3] Compiling Bar` line. I think there's a bug in the glob handling.

JordanMartinez commented:
The above issue was due to `z/src/**/*.purs` being expanded by the shell to `z/src/Bar/Baz.purs` before the compiler ever received the glob. Wrapping that in single quotes disables that behavior, such that the compiler receives the full glob correctly.

JordanMartinez commented:
This PR is ready for an initial review now.

However, I noticed an issue somewhat related to this PR with how we produce our globs. In #4480, we added the option to exclude files from the initial set of files produced by the globs. However, this change was only done in the `compile` command. So, other commands that use globs like `docs`, `graph`, `hierarchy`, and `repl` don't support the `exclude-files` flag.

Moreover, should the format of the glob file added here also include some code for the `excldue-files` globs, such that one could write something like the following where lines prefixed with `+` act like our current input globs and `-` act like the `exclude-files` arg?
```
+ src/**/*.purs
- src/Spec/**/*.purs
```

f-f commented:
> Moreover, should the format of the glob file added here also include some code for the excldue-files globs, such that one could write something like the following where lines prefixed with + act like our current input globs and - act like the exclude-files arg?

I think we should ship the feature as implemented, and we can issue a future patch if there's the need for this. I don't think the `exclude-files` feature is in high usage at the moment.

f-f reviewed (commented):
The patch is good, but what about the `glob-test.sh` file? We should either use it from a test suite, or remove it since it will rot if not machine-checked regularly.

JordanMartinez commented:
> The patch is good, but what about the `glob-test.sh` file? We should either use it from a test suite, or remove it since it will rot if not machine-checked regularly.

I've added it to our tests, but made it only run on Linux.

JordanMartinez commented:
> However, I noticed an issue somewhat related to this PR with how we produce our globs. In https://github.com/purescript/purescript/pull/4480, we added the option to exclude files from the initial set of files produced by the globs. However, this change was only done in the compile command. So, other commands that use globs like docs, graph, hierarchy, and repl don't support the exclude-files flag.

I've updated this PR to also add `--exclude-files` and `--source-globs` to more commands.

JordanMartinez commented:
Now that CI builds, can this get another review due to adding both `--source-globs` and `--exclude-files` to the other commands?

f-f commented on CHANGELOG.d/feature_glob-input-files.md:
````diff
@@ -0,0 +1,34 @@
+  purs compile --source-globs .spago/source-globs src/**/*.purs 
+  ```
+
+  In the command, `purs compile glob1 glob2 --source-globs globsFoundInFile --exclude-file glob3`,
````
Typo
```suggestion
  In the command, `purs compile glob1 glob2 --source-globs globsFoundInFile --exclude-files glob3`,
```

f-f commented on app/Command/Compile.hs:
```diff
@@ -72,25 +76,19 @@ compile PSCMakeOptions{..} = do
-excludedFiles = Opts.strOption $
+globInputFile :: Opts.Parser (Maybe FilePath)
+globInputFile = Opts.optional $ Opts.strOption $
+     Opts.long "source-globs"
```
I think the `exclude-files` is a bit misnamed now that we have a feature that lets us specify a file to include globs (feels like it should be `exclude-globs`?) - anyhow I wonder if this new flag should be called `source-globs-file` instead, to note that it's a file containing globs, and not just globs.
```suggestion
     Opts.long "source-globs-file"
```

JordanMartinez replied:
Personally, I think that would be a good change to make, but it's breaking. We could do it in the `0.16.0` release.

JordanMartinez replied:
Then again... by updating the meta var from 'FILE' to 'GLOB' I think it does make more sense. So, maybe this change isn't needed?

f-f commented on app/Command/Docs.hs:
```diff
@@ -132,6 +146,19 @@ pscDocsOptions = PSCDocsOptions <$> format <*> output <*> compileOutputDir <*> m
   inputFile = Opts.strArgument $
        Opts.metavar "FILE"
     <> Opts.help "The input .purs file(s)"
+  
+  globInputFile :: Opts.Parser (Maybe FilePath)
+  globInputFile = Opts.optional $ Opts.strOption $
+      Opts.long "source-globs"
+    <> Opts.metavar "FILE"
+    <> Opts.help "A file containing a line-separated list of .purs globs."
+
+  excludeFile :: Opts.Parser FilePath
+  excludeFile = Opts.strOption $
+      Opts.short 'x'
+    <> Opts.long "exclude-files"
+    <> Opts.help "Glob of .purs files to exclude from the supplied files."
+
```
These are theoretically separate flags for separate commands, but they really are the same flag and we're duplicating this code 5 times - feels like they should be shared?

JordanMartinez commented:
Anything else? Or is this good to merge?

rhendric commented on app/Command/Ide.hs:
```diff
@@ -57,11 +58,13 @@ listenOnLocalhost port = do
-  , _serverOutputPath :: FilePath
-  , _serverPort       :: Network.PortNumber
-  , _serverLoglevel   :: IdeLogLevel
+  { _serverDirectory     :: Maybe FilePath
+  , _serverGlobs         :: [FilePath]
+  , _serverGlobsFromFile :: Maybe FilePath
+  , _serverGlobsExcluded :: [FilePath]
+  , _serverOutputPath    :: FilePath
+  , _serverPort          :: Network.PortNumber
+  , _serverLoglevel      :: IdeLogLevel
```
```suggestion
  { _serverDirectory  :: Maybe FilePath
  , _serverGlobs      :: [FilePath]
  , _serverGlobsFromFile :: Maybe FilePath
  , _serverGlobsExcluded :: [FilePath]
  , _serverOutputPath :: FilePath
  , _serverPort       :: Network.PortNumber
  , _serverLoglevel   :: IdeLogLevel
```

rhendric commented on CHANGELOG.d/bug_add-exclude-file-to-more-commands.md:
```diff
@@ -0,0 +1,5 @@
+* Add `--exclude-file` to more commands
```
`bug_` isn't one of the sacred file name prefixes; you'll want `fix_`. (But maybe this is actually `feature_`?)

JordanMartinez replied:
I think this is the second time I've done this: calling it `bug_` rather than `fix_`. :smile: 
I've renamed it to `feature_`.

rhendric reviewed (approved):
Just a bunch of nits; merge once you've dealt with the ones you think are worth dealing with.

f-f commented on CHANGELOG.d/feature_glob-input-files.md:
````diff
@@ -0,0 +1,38 @@
+
+  In the command...
+  ```
+  purs compile inputGlob1 inputGlob2 --source-globs-file inputGlobsFoundInFile --exclude-files excludeGlob1
````
(github suggestions break in the presence of ```, so I'll comment line by line)

This is a bit of a nit, but I would rephrase this slightly:
```suggestion
  purs compile inputGlob1 inputGlob2 --source-globs-file fileWithMoreGlobs --exclude-files excludeGlob1
```

f-f commented on CHANGELOG.d/feature_glob-input-files.md:
```````diff
@@ -0,0 +1,38 @@
+  In the command...
+  ```
+  purs compile inputGlob1 inputGlob2 --source-globs-file inputGlobsFoundInFile --exclude-files excludeGlob1
+  ``````
```````
typo here?

There should be only three `

f-f reviewed (commented):
I left one nit in the changelog, but otherwise it's good to go - thanks @JordanMartinez!

JordanMartinez commented:
> I left one nit in the changelog, but otherwise it's good to go.

Fixed!
